pkg/query-service/app/traces/v3/query_builder.go

```diff
@@ -0,0 +1,385 @@
+	if mq.AggregateAttribute.Key != "" {
+		aggregationKey, err = getColumnName(mq.AggregateAttribute, keys)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	switch mq.AggregateOperator {
+	case v3.AggregateOperatorRateSum,
+		v3.AggregateOperatorRateMax,
+		v3.AggregateOperatorRateAvg,
+		v3.AggregateOperatorRateMin,
+		v3.AggregateOperatorRate:
+		op := fmt.Sprintf("%s(%s)/%d", aggregateOperatorToSQLFunc[mq.AggregateOperator], aggregationKey, step)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, having, orderBy)
+		return query, nil
+	case
+		v3.AggregateOperatorP05,
+		v3.AggregateOperatorP10,
+		v3.AggregateOperatorP20,
+		v3.AggregateOperatorP25,
+		v3.AggregateOperatorP50,
+		v3.AggregateOperatorP75,
+		v3.AggregateOperatorP90,
+		v3.AggregateOperatorP95,
+		v3.AggregateOperatorP99:
+		op := fmt.Sprintf("quantile(%v)(%s)", aggregateOperatorToPercentile[mq.AggregateOperator], aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, having, orderBy)
+		return query, nil
+	case v3.AggregateOperatorAvg, v3.AggregateOperatorSum, v3.AggregateOperatorMin, v3.AggregateOperatorMax:
+		op := fmt.Sprintf("%s(%s)", aggregateOperatorToSQLFunc[mq.AggregateOperator], aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, having, orderBy)
+		return query, nil
+	case v3.AggregateOperatorCount:
+		if mq.AggregateAttribute.Key != "" {
+			key, err := enrichKeyWithMetadata(mq.AggregateAttribute, keys)
+			if err != nil {
+				return "", err
+			}
+			columnType, columnDataType := getClickhouseTracesColumnDataTypeAndType(key)
+			filterSubQuery = fmt.Sprintf("%s AND has(%s%s, '%s')", filterSubQuery, columnDataType, columnType, mq.AggregateAttribute.Key)
+		}
+		op := "toFloat64(count())"
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, having, orderBy)
+		return query, nil
+	case v3.AggregateOperatorCountDistinct:
+		op := fmt.Sprintf("toFloat64(count(distinct(%s)))", aggregationKey)
+		query := fmt.Sprintf(queryTmpl, step, op, filterSubQuery, groupBy, having, orderBy)
+		fmt.Println(query)
+		return query, nil
+	case v3.AggregateOperatorNoOp:
+		// queryTmpl := constants.TracesSQLSelect + "from " + constants.SIGNOZ_TRACE_DBNAME + "." + constants.SIGNOZ_SPAN_INDEX_TABLENAME + " where %s %s"
+		// query := fmt.Sprintf(queryTmpl, spanIndexTableTimeFilter, filterSubQuery)
+		// return query, nil
+		return "", fmt.Errorf("not implemented, part of traces page")
+	default:
+		return "", fmt.Errorf("unsupported aggregate operator")
+	}
+}
+
+// groupBy returns a string of comma separated tags for group by clause
+// `ts` is always added to the group by clause
+func groupBy(tags ...string) string {
+	tags = append(tags, "ts")
+	return strings.Join(tags, ",")
+}
+
+func groupByAttributeKeyTags(keys map[string]v3.AttributeKey, tags ...v3.AttributeKey) (string, error) {
+	groupTags := []string{}
+	for _, tag := range tags {
+		groupTag, err := getColumnName(tag, keys)
```

Comment:
Good point!
This is happening because the map values are not nullable. Adding a where clause to check if the key exists can solve this.
Eg: Add `has(stringTagMap, 'key')` to where clause

@nityanandagohain This problem might also be there in logs, please check

Reply:
Solved